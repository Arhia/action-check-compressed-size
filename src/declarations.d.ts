declare module 'size-plugin-core'
