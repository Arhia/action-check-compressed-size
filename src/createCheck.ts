import { getOctokit } from '@actions/github'
import { Context } from '@actions/github/lib/context'
type Octokit = ReturnType<typeof getOctokit>

interface DetailsCheck {
    conclusion:
        | 'success'
        | 'failure'
        | 'neutral'
        | 'cancelled'
        | 'skipped'
        | 'timed_out'
        | 'action_required'
        | undefined
    output: {
        title: string
        summary: string
    }
}
/*
 * create a check and return a function that updates(completes) it
 */
export async function createCheck(octokit: Octokit, context: Context) {
    const check = await octokit.rest.checks.create({
        ...context.repo,
        name: 'Compressed Size',
        head_sha: context.payload.pull_request!.head.sha,
        status: 'in_progress'
    })

    return async (details: DetailsCheck) => {
        await octokit.rest.checks.update({
            ...context.repo,
            check_run_id: check.data.id,
            completed_at: new Date().toISOString(),
            status: 'completed',
            ...details
        })
    }
}
