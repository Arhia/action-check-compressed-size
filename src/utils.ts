import fs from 'fs'
import prettyBytes from 'pretty-bytes'

/**
 * Check if a given file exists and can be accessed.
 * @param {string} filename
 */
export async function fileExists(filename: string): Promise<boolean> {
    try {
        await fs.promises.access(filename, fs.constants.F_OK)
        return true
    } catch (e) {
        return false
    }
}

/**
 * Remove any matched hash patterns from a filename string.
 * @returns {(((fileName: string) => string) | undefined)}
 */
export function stripHash(allRegex: string[]): ((fileName: string) => string) | undefined {
    if (allRegex && allRegex.length) {
        return function (fileName: string): string {
            let finalStr = fileName
            for (const regex of allRegex) {
                finalStr = finalStr.replace(new RegExp(regex), (str, ...hashes) => {
                    hashes = hashes.slice(0, -2).filter(c => c != null)
                    if (hashes.length) {
                        for (const hash of hashes) {
                            const hashFormatted = hash || ''
                            str = str.replace(hashFormatted, hashFormatted.replace(/./g, '*'))
                        }
                        return str
                    }
                    return ''
                })
            }
            return finalStr
        }
    }

    return undefined
}

export function getDeltaText(delta: number, difference: number): string {
    let deltaText = (delta > 0 ? '+' : '') + prettyBytes(delta)
    if (delta && Math.abs(delta) > 1) {
        deltaText += ` (${Math.abs(difference)}%)`
    }
    return deltaText
}

export function iconForDifference(difference: number): string {
    let icon = ''
    if (difference >= 50) icon = 'üÜò'
    else if (difference >= 20) icon = 'üö®'
    else if (difference >= 10) icon = '‚ö†Ô∏è'
    else if (difference >= 5) icon = 'üîç'
    else if (difference <= -50) icon = 'üèÜ'
    else if (difference <= -20) icon = 'üéâ'
    else if (difference <= -10) icon = 'üëè'
    else if (difference <= -5) icon = '‚úÖ'
    return icon
}

/**
 * Create a Markdown table from text rows
 */
function markdownTable(rows: string[][]): string {
    if (!rows.length) {
        return ''
    }

    // Skip all empty columns
    while (rows.every(columns => !columns[columns.length - 1])) {
        for (const columns of rows) {
            columns.pop()
        }
    }

    const [firstRow] = rows
    const columnLength = firstRow.length
    if (columnLength === 0) {
        return ''
    }

    return [
        // Header
        ['Filename', 'Size', 'Change', ''].slice(0, columnLength),
        // Align
        [':---', ':---:', ':---:', ':---:'].slice(0, columnLength),
        // Body
        ...rows
    ]
        .map(columns => `| ${columns.join(' | ')} |`)
        .join('\n')
}

interface FileDiff {
    filename: string
    size: number
    delta: number
}

interface DiffOptions {
    showTotal: boolean
    collapseUnchanged: boolean
    omitUnchanged: boolean
    minimumChangeThreshold: number
}

/**
 * Create a Markdown table showing diff data
 * @param {Diff[]} files
 * @param {object} options
 * @param {boolean} [options.showTotal]
 * @param {boolean} [options.collapseUnchanged]
 * @param {boolean} [options.omitUnchanged]
 * @param {number} [options.minimumChangeThreshold]
 */
export function diffTable(
    files: FileDiff[],
    { showTotal, collapseUnchanged, omitUnchanged, minimumChangeThreshold }: DiffOptions
) {
    const changedRows = []
    const unChangedRows = []

    let totalSize = 0
    let totalDelta = 0
    for (const file of files) {
        const { filename, size, delta } = file
        totalSize += size
        totalDelta += delta

        const difference = ((delta / size) * 100) | 0
        const isUnchanged = Math.abs(delta) < minimumChangeThreshold

        if (isUnchanged && omitUnchanged) continue

        const columns = [
            `\`${filename}\``,
            prettyBytes(size),
            getDeltaText(delta, difference),
            iconForDifference(difference)
        ]
        if (isUnchanged && collapseUnchanged) {
            unChangedRows.push(columns)
        } else {
            changedRows.push(columns)
        }
    }

    let out = markdownTable(changedRows)

    if (unChangedRows.length !== 0) {
        const outUnchanged = markdownTable(unChangedRows)
        out += `\n\n<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>\n\n${outUnchanged}\n\n</details>\n\n`
    }

    if (showTotal) {
        const totalDifference = ((totalDelta / totalSize) * 100) | 0
        const totalDeltaText = getDeltaText(totalDelta, totalDifference)
        const totalIcon = iconForDifference(totalDifference)
        out = `**Total Size:** ${prettyBytes(totalSize)}\n\n${out}`
        out = `**Size Change:** ${totalDeltaText} ${totalIcon}\n\n${out}`
    }

    return out
}

/**
 * Convert a string "true"/"yes"/"1" argument value to a boolean
 */
export function toBool(v: string): boolean {
    return /^(1|true|yes)$/.test(v)
}
