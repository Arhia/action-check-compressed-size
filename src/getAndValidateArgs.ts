import { getInput } from '@actions/core'
import { toBool } from './utils'

type Args = {
    repoToken: string
    directory: string
    compression: string
    pattern: string
    exclude: string
    stripHashPattern: string[]
    minimumChangeThreshold: number
    showTotal: boolean
    omitUnchanged: boolean
    collapseUnchanged: boolean
}

export function getAndValidateArgs(): Args {
    const args = {
        repoToken: getInput('repo-token', { required: true }),
        directory: getInput('directory'),
        pattern: getInput('pattern') || '**/dist/**/*.js',
        exclude: getInput('exclude') || '{**/*.map,**/node_modules/**}',
        compression: getInput('compression'),
        stripHashPattern: getInput('strip-hash') ? JSON.parse(getInput('strip-hash')) : [],
        minimumChangeThreshold: parseInt(getInput('minimum-change-threshold'), 10),
        showTotal: toBool(getInput('show-total')),
        omitUnchanged: toBool(getInput('omit-unchanged')),
        collapseUnchanged: toBool(getInput('collapse-unchanged'))
    }

    return args
}
