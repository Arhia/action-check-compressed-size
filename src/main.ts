import * as core from '@actions/core'
import * as github from '@actions/github'
import { GitHub } from '@actions/github/lib/utils'
import * as micromatch from 'micromatch'
type GithubClient = InstanceType<typeof GitHub>
interface GithubIssue {
  body: string
  number: number
  active_lock_reason: string | null
  assignee: string | null
  assignees: string[]
  author_association: string
  closed_at: string | null
  created_at: string | null
  comments: number
  id: number
  title: string
  labels: string[]
}

type Args = {
  repoToken: string
  configPath: string
}

export type TriageBotConfig = {
  labels: {
    label: string
    glob: string
    comment?: string
    negate?: boolean
  }[]
  comment?: string
  no_label_comment?: string
}

async function run(): Promise<void> {
  try {
    const args = getAndValidateArgs()
    core.info('Starting GitHub Client')
    const client = github.getOctokit(args.repoToken)

    const issue = (github.context.payload.issue as unknown) as GithubIssue
    if (!issue) {
      core.error('No issue context found. This action can only run on issue creation.')
      return
    }
    core.debug(`Issue body content from context : \n ${issue.body}`)

    core.info(`Loading config file at ${args.configPath}`)
    const config = await getConfig(client, args.configPath)

    const { matchingLabels, comments } = processIssue({ config, issue })

    if (matchingLabels.length > 0) {
      core.info(`Adding labels ${matchingLabels.join(', ')} to issue #${issue.number}`)

      await addLabels(client, issue.number, matchingLabels)

      if (comments.length) {
        await writeComment(client, issue.number, comments.join('\n\n'))
      }
    } else if (config.no_label_comment) {
      core.info(`Adding comment to issue #${issue.number}, because no labels match`)

      await writeComment(client, issue.number, config.no_label_comment)
    }
  } catch (error) {
    core.error(error)
    core.setFailed(error.message)
  }
}

export function processIssue({
  config,
  issue
}: {
  config: TriageBotConfig
  issue: { body: string }
}): {
  matchingLabels: string[]
  comments: string[]
} {
  const matchingLabels: string[] = []
  const comments: string[] = config.comment ? [config.comment] : []

  const lines = issue.body.split(/\r?\n|\r/g)

  for (const label of config.labels) {
    const isNegated = label.negate === true
    let isMatching: boolean
    if (isNegated) {
      // si on negate le patern, aucune ligne ne doit contenir le pattern
      isMatching = !lines.some(l => micromatch.isMatch(l, label.glob))
    } else {
      // pattern normal, au mois 1 ligne doit contenir le pattern
      isMatching = lines.some(l => micromatch.isMatch(l, label.glob))
    }

    if (isMatching) {
      core.debug(`Match in body for pattern ${label.glob}`)
      matchingLabels.push(label.label)
      if (label.comment) {
        comments.push(label.comment)
      }
    } else {
      core.debug(`No match in body for pattern ${label.glob}`)
    }
  }
  return {
    matchingLabels,
    comments
  }
}

async function writeComment(client: GithubClient, issueId: number, body: string): Promise<void> {
  await client.issues.createComment({
    owner: github.context.repo.owner,
    repo: github.context.repo.repo,
    issue_number: issueId,
    body
  })
}

async function addLabels(client: GithubClient, issueId: number, labels: string[]): Promise<void> {
  await client.issues.addLabels({
    owner: github.context.repo.owner,
    repo: github.context.repo.repo,
    issue_number: issueId,
    labels
  })
}

async function getConfig(client: GithubClient, configPath: string): Promise<TriageBotConfig> {
  const response = await client.repos.getContent({
    owner: github.context.repo.owner,
    repo: github.context.repo.repo,
    path: configPath,
    ref: github.context.sha
  })

  return JSON.parse(Buffer.from(response.data.content, 'base64').toString())
}

function getAndValidateArgs(): Args {
  const args = {
    repoToken: core.getInput('repo-token', { required: true }),
    configPath: core.getInput('config-path', { required: true })
  }

  return args
}

run()
