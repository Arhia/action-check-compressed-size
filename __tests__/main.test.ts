import { TriageBotConfig, processIssue } from '../src/main'

test('1.', () => {
  const config: TriageBotConfig = {
    labels: [
      {
        label: ':warning:Status: invalid (issue template not used)',
        glob: '*GENERATED_BY_TEMPLATE*',
        negate: true,
        comment:
          "Bonjour\nVous n'avez pas utilisé le modèle de création de ticket, votre ticket ne pourra pas être traité.\n\n Merci de sélectionner le modèle approprié à la création du ticket quand vous cliquez sur 'New Issue' "
      }
    ]
  }
  const body = 'zerzer\r\n\r\nzer\r\n\r\nzer\r\n\r\n\r\nGENERATED_BY_TEMPLATE\r\n'
  const { matchingLabels, comments } = processIssue({
    config,
    issue: { body }
  })
  expect(matchingLabels.length).toEqual(0)
  expect(comments.length).toEqual(0)
})

test('2.', () => {
  const config: TriageBotConfig = {
    labels: [
      {
        label: ':warning:Status: invalid (issue template not used)',
        glob: '*GENERATED_BY_TEMPLATE*',
        negate: true,
        comment:
          "Bonjour\nVous n'avez pas utilisé le modèle de création de ticket, votre ticket ne pourra pas être traité.\n\n Merci de sélectionner le modèle approprié à la création du ticket quand vous cliquez sur 'New Issue' "
      }
    ]
  }
  const body = 'zerzer\r\n\r\nzer\r\n\r\nzer\r\n\r\n\reee\r\n'
  const { matchingLabels, comments } = processIssue({
    config,
    issue: { body }
  })
  expect(matchingLabels.length).toEqual(1)
  expect(comments.length).toEqual(1)
})

test('3.', () => {
  const config: TriageBotConfig = {
    labels: [
      {
        label: ':warning:Status: invalid (issue template not used)',
        glob: '*GENERATED_BY_TEMPLATE*',
        negate: true,
        comment:
          "Bonjour\nVous n'avez pas utilisé le modèle de création de ticket, votre ticket ne pourra pas être traité.\n\n Merci de sélectionner le modèle approprié à la création du ticket quand vous cliquez sur 'New Issue' "
      }
    ]
  }
  const body = `
  @
2020-08-25T10:44:26.2817988Z @
2020-08-25T10:44:26.2818174Z
2020-08-25T10:44:26.2819412Z ## Résultat observé
2020-08-25T10:44:26.2819601Z
2020-08-25T10:44:26.2820329Z Le Bloc paiement est impossible à supprimer alors que la DSN associée est supprimée
2020-08-25T10:44:26.2820551Z
2020-08-25T10:44:26.2821485Z ![image](https://user-images.githubusercontent.com/66441178/91160764-1ba80580-e6ca-11ea-9348-0a525038b862.png)
2020-08-25T10:44:26.2821811Z
2020-08-25T10:44:26.2822349Z <!--
2020-08-25T10:44:26.2822521Z
2020-08-25T10:44:26.2823034Z -->
2020-08-25T10:44:26.2823201Z
2020-08-25T10:44:26.2823757Z ## Résultat attendu
2020-08-25T10:44:26.2824042Z
2020-08-25T10:44:26.2824715Z Possibilité de supprimer le bloc paiement si la DSN est supprimée.
2020-08-25T10:44:26.2825272Z <!--
2020-08-25T10:44:26.2825440Z
2020-08-25T10:44:26.2825952Z -->
2020-08-25T10:44:26.2826143Z
2020-08-25T10:44:26.2826736Z ## Comment reproduire le problème ?
2020-08-25T10:44:26.2826924Z
2020-08-25T10:44:26.2827439Z <!--
2020-08-25T10:44:26.2827760Z pas sure :
2020-08-25T10:44:26.2828326Z cloturer la paie d'Aout,
2020-08-25T10:44:26.2831097Z lancer la DSN de Septembre (erreur) au lieu de celle du mois d'aout (car le logiciel propose par défaut d'effectuer la DSN du mois suivant non cloturé).
2020-08-25T10:44:26.2832057Z Supprimer la DSN de Septembre, mais laisser le bloc paiement associé (oubli).
2020-08-25T10:44:26.2832686Z Faire la DSN d'aout.
2020-08-25T10:44:26.2833024Z Faire la paie de Septembre
2020-08-25T10:44:26.2833386Z Essayer de supprimer le bloc paiement de septembre pour faire la DSN de Septembre.
2020-08-25T10:44:26.2834009Z -->
2020-08-25T10:44:26.2834215Z
2020-08-25T10:44:26.2834776Z - Dossier : PERRIER
2020-08-25T10:44:26.2835339Z - Période : Aout
2020-08-25T10:44:26.2835899Z - Salarié :  NA
2020-08-25T10:44:26.2836592Z - Action à faire par l'utilisateur pour déclencher le résultat observé :
2020-08-25T10:44:26.2837733Z pas sure : cloturer la paie d'Aout, lancer la DSN de Septembre (erreur) au lieu de celle du mois d'aout (car le logiciel propose par défaut d'effectuer la DSN du mois suivant non cloturé). Supprimer la DSN de Septembre, mais laisser le bloc paiement associé.
2020-08-25T10:44:26.2838423Z Faire la DSN d'aout.
2020-08-25T10:44:26.2838771Z Faire la paie de Septembre
2020-08-25T10:44:26.2839968Z Essayer de supprimer le bloc paiement de septembre pour faire la DSN de Septembre.
2020-08-25T10:44:26.2840189Z
2020-08-25T10:44:26.2840899Z ## Le cas échéant : analyse et piste de résolution
2020-08-25T10:44:26.2841793Z Que le logiciel ne propose pas de faire la DSN du mois ouvert pour la paie, mais la DSN du mois qui n'a pas encore été réalisée.
2020-08-25T10:44:26.2843041Z En effet, il arrive que le client ait besoin d'accéder au mois suivant pour saisir sa paie alors que l'on a pas terminé les DSN.
2020-08-25T10:44:26.2843279Z
2020-08-25T10:44:26.2843810Z <!--
2020-08-25T10:44:26.2844583Z Ne renseignez cette partie que si vous pensez connaitre l'origine technique du problème, sinon laisser vide*
2020-08-25T10:44:26.2845304Z Dans tous les cas, une analyse sera effectuée par le service R&D*
2020-08-25T10:44:26.2845860Z -->
2020-08-25T10:44:26.2846029Z
2020-08-25T10:44:26.2846216Z
2020-08-25T10:44:26.2846856Z <!-- GENERATED_BY_TEMPLATE. NE PAS EFFACER CETTE LIGNE -->
2020-08-25T10:44:26.2847056Z
`
  const { matchingLabels, comments } = processIssue({
    config,
    issue: { body }
  })
  expect(matchingLabels.length).toEqual(0)
  expect(comments.length).toEqual(0)
})
