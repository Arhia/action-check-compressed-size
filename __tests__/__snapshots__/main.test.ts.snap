// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffTable 1`] = `
"**Size Change:** +9 B (0%) 

**Total Size:** 4.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (50%) | üÜò |
| \`two.js\` | -5 kB | -2.5 kB (50%) | üÜò |
| \`four.js\` | 4.5 kB | +9 B (0%) |  |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`three.js\` | 300 B | 0 B |

</details>

"
`;

exports[`diffTable 2`] = `
"| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (50%) | üÜò |
| \`two.js\` | -5 kB | -2.5 kB (50%) | üÜò |
| \`four.js\` | 4.5 kB | +9 B (0%) |  |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`three.js\` | 300 B | 0 B |

</details>

"
`;

exports[`diffTable 3`] = `
"**Size Change:** +9 B (0%) 

**Total Size:** 4.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (50%) | üÜò |
| \`two.js\` | -5 kB | -2.5 kB (50%) | üÜò |
| \`three.js\` | 300 B | 0 B |  |
| \`four.js\` | 4.5 kB | +9 B (0%) |  |"
`;

exports[`diffTable 4`] = `
"**Size Change:** +9 B (0%) 

**Total Size:** 4.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (50%) | üÜò |
| \`two.js\` | -5 kB | -2.5 kB (50%) | üÜò |
| \`four.js\` | 4.5 kB | +9 B (0%) |  |"
`;

exports[`diffTable 5`] = `
"**Size Change:** +9 B (0%) 

**Total Size:** 4.8 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| \`one.js\` | 5 kB | +2.5 kB (50%) | üÜò |
| \`two.js\` | -5 kB | -2.5 kB (50%) | üÜò |

<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`three.js\` | 300 B | 0 B |
| \`four.js\` | 4.5 kB | +9 B (0%) |

</details>

"
`;

exports[`diffTable 6`] = `
"**Size Change:** 0 B 

**Total Size:** 4.8 kB



<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`one.js\` | 5 kB | 0 B |
| \`two.js\` | -5 kB | 0 B |
| \`three.js\` | 300 B | 0 B |
| \`four.js\` | 4.5 kB | 0 B |

</details>

"
`;

exports[`diffTable 7`] = `
"**Size Change:** 0 B 

**Total Size:** 300 B



<details><summary>‚ÑπÔ∏è <strong>View Unchanged</strong></summary>

| Filename | Size | Change |
| :--- | :---: | :---: |
| \`three.js\` | 300 B | 0 B |

</details>

"
`;
